### ISSUE TITLE
[MaterialTimePicker] Add Input validation

### ISSUE THREAD

**Issue by yashovardhan99** [material-components/material-components-android#2106]
**Is your feature request related to a problem? Please describe.**
Many times, apps need users to select a date+time in the past/future. `MaterialDatePicker` solves this problem by having the option to add constraints with a validator. TimePicker doesn't allow any such option.

**Describe the solution you'd like**
An `addConstraints` implementation for `MaterialTimePicker`. I understand that it won't be possible to simply extend `CalendarConstraints` for this because the time picker is only concerned with time but an option to add constraints in the form of (say) `constraintBefore(int hour, int minute)` and `constraintAfter(int hour, int minute)` can be added.

**Describe alternatives you've considered**
The present alternative to this I am using is to manually check the time from the `TimePicker` and reset the time if it doesn't fall within the required constraints. (Say I only allow time to be selected in the past, but user has entered a future time. I reset the displayed time to the current time). But this leaves room for a bad UX where the user can get confused or not even realize their selected time has been changed.

**Additional context**
An example implementation can be [this library I found on Github](https://github.com/wdullaer/MaterialDateTimePicker#timepickerdialog-setmintimetimepoint-time)

**Comment by drchen** [comment]
I guess the concern here will be, what should be the UX feedback when an invalid time is selected? Also we cannot block invalid time from being selected due to the three-step selection of time (hour -> minute -> am/pm). 

Maybe the best we can do is disable the ok button when the time is invalid, will that work in your scenario?

**Comment by yashovardhan99** [comment]
Yeah that's a valid point. I guess disabling the ok button could be a good starting point.

**Comment by yashovardhan99** [comment]
> Also we cannot block invalid time from being selected due to the three-step selection of time (hour -> minute -> am/pm).

How about disabling specific hours in 24 hours and disabling AM/PM depending on hour selected for 12 hours? This can then be extended to disabling specific minutes based on the hour selected as well? I know it is easier said then done but I just want to know if you think this can be a viable alternative?

To give [an example](https://github.com/wdullaer/MaterialDateTimePicker#timepickerdialog-setmintimetimepoint-time), the library I quoted adds the following methods:-
```
setMinTime()
setMaxTime()
setSelectableTimes()
setDisabledTimes()
```

I am not sure how much of a change will be required to make this viable though?

**Comment by drchen** [comment]
Hi I think disabling OK button is feasible since it has less UX impact. Regarding your suggestion - we may need to consult with our UX designers and see what they think, which, may take a much longer time. 

I'll raise the idea internally first and check if we can add support to disable OK button first.

**Comment by BillyLuisNeedham** [comment]
Is there any progress on this? We have also an issue where we would like to apply constraints.

**Comment by Red7M** [comment]
Just so we know, are time constraints a "in progress" work currently?

**Comment by theblakdrag0n** [comment]
> I guess the concern here will be, what should be the UX feedback when an invalid time is selected? Also we cannot block invalid time from being selected due to the three-step selection of time (hour -> minute -> am/pm).
> 
> Maybe the best we can do is disable the ok button when the time is invalid, will that work in your scenario?

drchen, I have a situation at work where we need to restrict the user selectable time.  I would like to know why you say you cannot block selection in the three-step (Clock mode) selection process?  As yashovardhan99 mentioned, it seems to me that since the Date Picker greys out and disables selection of restricted days that you could do exactly the same thing for the Time Picker clock style picker hours, minutes, and AM/PM selection?

I would love to see Time Picker restrictions implemented just like the Date Picker (where unselectable values are disabled/not clickabe) because once this is implemented I would be able to use the TIme Picker in a project at work.

**Comment by michael-winkler** [comment]
@drchen Any progress here? A min and max for hour and minutes would be great! 👍

**Comment by drchen** [comment]
Sorry that I haven't been actively working on Material3 Views for a while. Hunter can you take over this and estimate if we can implement this?
